src/main/java/it/gov/pagopa/payment/notice/generator/service/NoticeGenerationServiceImpl.java

```diff
@@ -0,0 +1,232 @@
+import java.util.Optional;
+
+import static it.gov.pagopa.payment.notice.generator.util.WorkingDirectoryUtils.clearTempDirectory;
+import static it.gov.pagopa.payment.notice.generator.util.WorkingDirectoryUtils.createWorkingDirectory;
+
+/**
+ * Services regarding the notice generation flow
+ */
+@Service
+@Slf4j
+public class NoticeGenerationServiceImpl implements NoticeGenerationService {
+
+    private final PaymentGenerationRequestErrorRepository paymentGenerationRequestErrorRepository;
+    private final PaymentGenerationRequestRepository paymentGenerationRequestRepository;
+
+    private final PdfEngineClient pdfEngineClient;
+
+    private final InstitutionsStorageClient institutionsStorageClient;
+    private final NoticeStorageClient noticeStorageClient;
+    private final NoticeTemplateStorageClient noticeTemplateStorageClient;
+
+    private final Aes256Utils aes256Utils;
+
+    private final ObjectMapper objectMapper;
+
+    private final Validator validator;
+
+    public NoticeGenerationServiceImpl(
+            PaymentGenerationRequestRepository paymentGenerationRequestRepository,
+            PaymentGenerationRequestErrorRepository paymentGenerationRequestErrorRepository,
+            InstitutionsStorageClient institutionsStorageClient,
+            NoticeStorageClient noticeStorageClient,
+            NoticeTemplateStorageClient noticeTemplateStorageClient,
+            PdfEngineClient pdfEngineClient,
+            Aes256Utils aes256Utils,
+            ObjectMapper objectMapper,
+            Validator validator) {
+        this.paymentGenerationRequestRepository = paymentGenerationRequestRepository;
+        this.paymentGenerationRequestErrorRepository = paymentGenerationRequestErrorRepository;
+        this.institutionsStorageClient = institutionsStorageClient;
+        this.noticeStorageClient = noticeStorageClient;
+        this.noticeTemplateStorageClient = noticeTemplateStorageClient;
+        this.pdfEngineClient = pdfEngineClient;
+        this.aes256Utils = aes256Utils;
+        this.objectMapper = objectMapper;
+        this.validator = validator;
+    }
+
+    /**
+     * Generate a notice, if required content is provided and valid, and saves the content to a folderId if provided
+     * and valid
+     * @param noticeGenerationRequestItem request data to use for the notice generation
+     * @param folderId optional parameter to generate folderId
+     * @return generated notice
+     */
+    @SneakyThrows
+    @Override
+    public File generateNotice(NoticeGenerationRequestItem noticeGenerationRequestItem,
+                               String folderId) {
+
+        if (folderId != null) {
+            Optional<PaymentNoticeGenerationRequest> paymentNoticeGenerationRequestOptional =
+                    paymentGenerationRequestRepository.findById(folderId);
+            if (paymentNoticeGenerationRequestOptional.isEmpty()) {
+                throw new AppException(AppError.FOLDER_NOT_AVAILABLE);
+            }
```

Comment:
we can use all the potential of streams

```suggestion
            PaymentNoticeGenerationRequest paymentNoticeGenerationRequestOptional =
                    paymentGenerationRequestRepository.findById(folderId)
                    .orElseThrow(()->throw new AppException(AppError.FOLDER_NOT_AVAILABLE));
```